concrete-csprng/src/seeders/implem/linux.rs

```diff
@@ -0,0 +1,59 @@
+use crate::seeders::{Seed, Seeder};
+use std::fs::File;
+use std::io::Read;
+
+/// A seeder which uses the linux `/dev/random` source.
+pub struct LinuxSeeder {
+    counter: u128,
+    secret: u128,
+    file: File,
+}
+
+impl LinuxSeeder {
+    /// Creates a new seeder from a user defined secret.
+    ///
+    /// Important:
+    /// ----------
+    ///
+    /// This secret is used to ensure the quality of the seed in scenarios where `/dev/random` may
+    /// be compromised.
+    pub fn new(secret: u128) -> LinuxSeeder {
+        let file = std::fs::File::open("/dev/random").expect("Failed to open /dev/random .");
+        let counter = std::time::UNIX_EPOCH
+            .elapsed()
+            .expect("Failed to initialized software rdseed counter.")
+            .as_nanos();
+        LinuxSeeder {
+            secret,
+            counter,
+            file,
+        }
+    }
+}
+
+impl Seeder for LinuxSeeder {
+    fn seed(&mut self) -> Seed {
+        let output = self.secret ^ self.counter ^ dev_random(&mut self.file);
+        self.counter = self.counter.wrapping_add(1);
```

Comment:
if dev_random is compromised it feels like all seeds are then predictable as soon as you have 2 seeds, may not be an issue if the seeder is used only once (though if dev_random is compromised maybe the unix_epoch elapsed is also compromised) making seeding completely determinate.

We would need some opinion on what scenario should be considered and what property the seeder should have ?

I really can't tell as it's outside of my current knowledge on PRNG and seeding

Reply:
This approach was proposed by a cryptographer here at zama. I merely implemented the suggested approach :sweat_smile: ! Maybe we could ask a cryptographer to have a look at this, and describe better the attack envisionned ?

Reply:
for sure as I said I really don't know what the attack would be ^^" and I don't know enough about PRNG and seeding